README


This archive contains the distribution Mail-MtPolicyd,
version 1.15:

  a modular policy daemon for postfix

This software is Copyright (c) 2014 by Markus Benning <ich@markusbenning.de>.

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991


This README file was generated by Dist::Zilla::Plugin::Readme v5.029.



CHANGES
==================================================
Changes from 2014-01-07 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 1.15 at 2015-01-07 15:21:18 +0000
-----------------------------------------

  Change: 2a46db07eb9f5a4a558355bdbadd9d6224c643dc
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2015-01-07 16:17:56 +0000

    Documentation updates

    * added how to for Accounting/Quotas

    * add section about cron() in Cookbook::ExtendedPlugin 

  Change: 635fc08b7b9ddc91a264966cf475be90a18345a4
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2015-01-07 12:50:33 +0000

    Added Quota plugin

    * new Plugin: Quota

    * make field for SQL configurable in SqlUserConfig

    * add tests for Quota to accounting test cases 

  Change: 08cf926d66643d371a36427dc01f49f7aa9f5461
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2015-01-06 19:35:39 +0000

    Added support for cron tasks

    * added a crontab to call mtpolicyd periodically

    * add mtpolicyd --cron <tasks> to execute tasks

    * plugins can do cron jobs by implementing a cron()

    * added autowl expire to Plugin::Greylist based on cron() 

  Change: 77ad175c055c2bc5dcb27ded0ae60cdcee13be6a
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-12-23 13:31:59 +0000

    Small fixes

    * initialize SQL connection before plugins

    * skip also emtpy or whitespace request attributes 

  Change: 059c1a4aae0001ba0b02044ce18e855f227cfaff
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-12-23 12:46:47 +0000

    Added Accounting Plugin 

  Change: 94380a398c18a79026f9abe8d849931e7387aed7
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-12-22 14:03:24 +0000

    More SQL improvements

    * add unit tests t/plugin-role-sqlutils.t t/sqlconnection.t

    * add support for automatic table creation

    Greylisting Plugin

    * fix autowl expire

    * add unit test for greylisting plugin 

  Change: cfa1805d549c5717b7abc3548e234ebd1505f792
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-12-22 09:15:59 +0000

    Add some SQL infrastructure

    * added Mail::MtPolicyd::SqlConnection singleton for storing DB

    connnection

    * added Plugin::Role::SqlUtils for code shared by plugins

    * adjust existing plugins

    * adjust tests 

  Change: 42a968ccd63f6cc1eb0c35bf64b3bb713a77095d
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-12-22 09:14:15 +0000

    Fix RBL test IP

    * use XBL test IP instead of a real client IP 

  Change: cefd767f93b66d259638e2d2d1c69801d27f8748
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-12-19 16:23:21 +0000

    Add cucumber test cases

    * added basic step files for cucumber test

    * add 2 basic server tests (minimal.conf and spamhaus-rbls.conf) 

  Change: 86eff808ab85a144d70f4442ac1199e940f53d86
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-12-19 15:30:43 +0000

    Dont fail in SPF when sender/client_address unavailable 

  Change: 8ebd94873931b78b9f86ce8909ebb22fe0a727cf
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-12-18 13:15:05 +0000

    fixed missing dependencies in debian package 

-----------------------------------------
version 1.14 at 2014-11-28 12:41:14 +0000
-----------------------------------------

  Change: 2a8b15a7e7c6dd764bf141ee5492c4fec3e7a207
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-11-28 13:41:14 +0000

    Small fixes

    * fix declaration of @plugin_results inside of eval in PluginChain

    * fix undefined $session in Plugin::Stress 

  Change: ee12431e845d84ede9e69bddeace3d53227e0b69
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-11-28 13:36:12 +0000

    Adjust documentation for on_error feature 

  Change: 86ca31eee7089ab78b317adacfd81ddef04eb06e
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-11-28 13:20:32 +0000

    Add on_error=continue feature to Plugins

    Added attribute on_error to Plugin base class. If set to continue and
    the plugin dies mtpolicyd will continue with processing instead of
    returning an error.

    <Plugin fault>

    module = "Faulty"

    on_error = "continue"

    </Plugin>

    If this plugin dies mtpolicyd will just log the plugins failure and
    continure. 

  Change: 52ca1f09675b97bf0b13b8bf727eb72bc9f233f7
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-11-20 14:02:08 +0000

    Small doc update

    use blacklist.db instead of whitelist.db in blacklist example. 

  Change: f3ceb5809399c7137e9017a990b4a4137d13909d
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-11-20 13:58:46 +0000

    New Plugin: Stress

    Added new plugin for postfix stress feature. 

  Change: 7a5508e97f15d714a8e23ca73b1eae463f2ddf2c
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-11-18 14:13:46 +0000

    Small fixes

    * fixed version string in dzil changelog generators

    * fix stop -> stop_current_timer 

  Change: afe9fc6bac1924c61d053d72354c123c63e8da18
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-11-17 10:32:22 +0000

    Added infrastructur for profiling

    This adds a application level profiling to mtpolicyd. Plugins may use
    this via Mail::MtPolicyd::Profiler like

    Mail::MtPolicyd::Profiler->tick('doing something');

    # do something

    Mail::MtPolicyd::Profiler->tick('finished something');

    Timings are logged at log level 3. 

  Change: 4b3b0c16891a6d25152b557167f7d54052996abe
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-09-15 15:03:19 +0000

    add honeypot config sample to default configuration 

  Change: e0d3d9af2fc4e2d8561b39b8f4e22dbe44348786
  Author: Markus Benning <ich@markusbenning.de>
  Date : 2014-09-15 07:27:54 +0000

    add .editorconfig file 

-----------------------------------------
version 1.13 at 2014-09-12 13:18:04 +0000

lib/Mail/MtPolicyd.pm
package Mail::MtPolicyd;

use strict;
use base qw(Net::Server::PreFork);

our $VERSION = '1.15'; # VERSION
# ABSTRACT: a modular policy daemon for postfix

use Data::Dumper;
use Mail::MtPolicyd::Profiler;
use Mail::MtPolicyd::Request;
use Mail::MtPolicyd::VirtualHost;
use Mail::MtPolicyd::SqlConnection;
use DBI;
use Cache::Memcached;
use Time::HiRes qw( usleep tv_interval gettimeofday );
use Getopt::Long;
use Tie::IxHash;
use Config::General qw(ParseConfig);
use IO::Handle;
 

sub _preload_modules {
	# PRELOAD some modules
	my @modules = (
		'BerkeleyDB',
		'BerkeleyDB::Hash',
		'DBI',
		'DBD::mysql',
		'HTTP::Request::Common',
		'JSON',
		'LWP::UserAgent',
		'Mail::RBL',
		'Moose',
		'Moose::Role',
		'MooseX::Getopt',
		'MooseX::Role::Parameterized',
		'namespace::autoclean',
	);

	foreach my $module (@modules) {
		$module =~ s/::/\//g;
		$module .= '.pm';
		require $module;
	}
}

sub _apply_values_from_config {
	my ( $self, $target, $config ) = ( shift, shift, shift );

	while ( my $key = shift ) {
		if(! defined $config->{$key} ) {
			next;
		}
		$target->{$key} = $config->{$key};
	}

	return;
}

sub _apply_array_from_config {
	my ( $self, $target, $config ) = ( shift, shift, shift );

	while ( my $key = shift ) {
		if(! defined $config->{$key} ) {
			next;
		}
		$target->{$key} = [ split(/\s*,\s*/, $config->{$key}) ];
	}

	return;
}

sub print_usage {
	print "mtpolicyd [-h|--help] [-c|--config=<file>] [-f|--foreground] [-l|--loglevel=<level>] [-d|--dump_vhosts] [-t|--cron=<task1,hourly,daily,...>]\n";
	return;
}

sub configure {
	my $self = shift;
	my $server = $self->{'server'};
	my $cmdline;
	
	return if(@_);

    if( ! defined $server->{'config_file'} ) {
   	    $server->{'config_file'} = '/etc/mtpolicyd/mtpolicyd.conf';
    }
	$server->{'background'} = 1;
	$server->{'setsid'} = 1;
	$server->{'no_close_by_child'} = 1;

        # Parse command line params
        %{$cmdline} = ();
        GetOptions(
            \%{$cmdline},
            "help|h",
            "dump_config|d",
            "config|c:s",
            "foreground|f",
            "loglevel|l:i",
            "cron|t:s",
        );
        if ($cmdline->{'help'}) {
                $self->print_usage;
                exit 0;
        }
        if (defined($cmdline->{'config'}) && $cmdline->{'config'} ne "") {
                $server->{'config_file'} = $cmdline->{'config'};
        }
	if( ! -f $server->{'config_file'} ) {
		print(STDERR 'configuration file '.$server->{'config_file'}.' does not exist!\n');
		exit 1;
	}

	# DEFAULTS
    if( ! defined $server->{'log_level'} ) {
	    $server->{'log_level'} = 2;
    }
	if( ! defined $server->{'log_file'} && ! $cmdline->{'foreground'} ) {
		$server->{'log_file'} = 'Sys::Syslog';
	}
	$server->{'syslog_ident'} = 'mtpolicyd';
	$server->{'syslog_facility'} = 'mail';

	$server->{'proto'} = 'tcp';
	$server->{'host'} = '127.0.0.1';
    if( ! defined $server->{'port'} ) {
    	$server->{'port'} = [ '127.0.0.1:12345' ];
    }

	$server->{'min_servers'} = 4;
        $server->{'min_spare_servers'} = 4;
        $server->{'max_spare_servers'} = 12;
        $server->{'max_servers'} = 25;
	$server->{'max_requests'} = 1000;

	$self->{'request_timeout'} = 20;

	$self->{'keepalive_timeout'} = 60;
	$self->{'max_keepalive'} = 0;

	$self->{'db_dsn'} = undef;
	$self->{'db_user'} = '';
	$self->{'db_password'} = '';

	$self->{'memcached_servers'} = [ '127.0.0.1:11211' ];
	$self->{'memcached_namespace'} = 'mt-';
	$self->{'memcached_expire'} = 5 * 60;

	# will be incremented in linear steps 50, 100, 150...
	$self->{'session_lock_wait'} = 50; # usec
	$self->{'session_lock_max_retry'} = 50; # times
	$self->{'session_lock_timeout'} = 10; # sec

	$self->{'program_name'} = $0;

	# APPLY values from configuration file
	tie my %config_hash, "Tie::IxHash";
	%config_hash = ParseConfig(
	  -AllowMultiOptions => 'no',
          -ConfigFile => $server->{'config_file'},
          -Tie => "Tie::IxHash"
        );
	my $config = \%config_hash;

	$self->_apply_values_from_config($server, $config, 
		'user', 'group', 'pid_file',
		'log_level', 'log_file', 'syslog_ident', 'syslog_facility',
		'host',
		'min_servers', 'min_spare_servers', 'max_spare_servers',
		'max_servers', 'max_requests',
		'chroot',
	);
	$self->_apply_array_from_config($server, $config, 'port');

	$self->_apply_values_from_config($self, $config, 
		'request_timeout', 'keepalive_timeout', 'max_keepalive',
		'db_dsn', 'db_user', 'db_password',
		'memcached_namespace', 'memcached_expire',
		'session_lock_wait', 'session_lock_max_retry', 'session_lock_timeout',
		'program_name',
	);
	$self->_apply_array_from_config($self, $config, 'memcached_servers');

    # Initialize DB connection before load vhosts
	if( defined $self->{'db_dsn'} && $self->{'db_dsn'} !~ /^\s*$/ ) {
        Mail::MtPolicyd::SqlConnection->initialize(
            dsn => $self->{'db_dsn'},
            user => $self->{'db_user'},
            password => $self->{'db_password'},
        );
	}

	# LOAD VirtualHosts
	if( ! defined $config->{'VirtualHost'} ) {
		print(STDERR 'no virtual hosts configured!\n');
		exit 1;
	}
	my $vhosts = $config->{'VirtualHost'};

	$self->{'virtual_hosts'} = {};
	foreach my $vhost_port (keys %$vhosts) {
		my $vhost = $vhosts->{$vhost_port};
		$self->{'virtual_hosts'}->{$vhost_port} = 
			Mail::MtPolicyd::VirtualHost->new_from_config($vhost_port, $vhost)
	}
    if ($cmdline->{'dump_config'}) {
        print "----- Virtual Hosts -----\n";
        print Dumper( $self->{'virtual_hosts'} );
        exit 0;
    }

	# foreground mode (cmdline)
    if ($cmdline->{'foreground'}) {
        $server->{'background'} = undef;
        $server->{'setsid'} = undef;
    }
	if( $cmdline->{'loglevel'} ) {
		$server->{'log_level'} = $cmdline->{'loglevel'};
	} 

    # if running in cron mode execute cronjobs and exit
    if( $cmdline->{'cron'} && $cmdline->{'cron'} !~ /^\s*$/ ) {
        my @tasks = split(/\s*,\s*/, $cmdline->{'cron'});
        $self->cron( @tasks );
        exit 0;
    }

	# change processname in top/ps
	$self->_set_process_stat('master');

	return;
}

sub cron {
    my $self = shift;
    foreach my $vhost ( keys %{$self->{'virtual_hosts'}} ) {
        $self->{'virtual_hosts'}->{$vhost}->cron( $self, @_ );
    }
    return;
}

sub pre_loop_hook {
	my $self = shift;

	$self->_preload_modules;

	return;
}

sub child_init_hook {
	my $self = shift;

	$self->_set_process_stat('virgin child');

    # close parent database connection
    if( Mail::MtPolicyd::SqlConnection->is_initialized ) {
        Mail::MtPolicyd::SqlConnection->disconnect;
    }

	$self->{'memcached'} = Cache::Memcached->new( {
		'servers' => $self->{'memcached_servers'},
		'debug' => 0,
		'namespace' => $self->{'memcached_namespace'},
	} );

	return;
}

sub child_finish_hook {
	my $self = shift;
	$self->_set_process_stat('finish');

	if( Mail::MtPolicyd::SqlConnection->is_initialized ) {
		eval { Mail::MtPolicyd::SqlConnection->instance->disconnect };
	}

	return;
}

sub memcached {
	my $self = shift;
	if( ! defined $self->{'memcached'} ) {
		die('no memcached connection available!');
	}
	return( $self->{'memcached'} );
}

sub acquire_session_lock {
	my ( $self, $instance ) = @_;
	my $lock = 'lock_'.$instance;
	my $wait = $self->{'session_lock_wait'};
	my $max_retry = $self->{'session_lock_max_retry'};
	my $lock_ttl = $self->{'session_lock_timeout'};

	for( my $try = 1 ; $try < $max_retry ; $try++ ) {
		if( $self->{'memcached'}->add($lock, 1, $lock_ttl) ) {
			return; # lock created
		}
		usleep( $wait * $try );
	}

	die('could not acquire lock for session '.$instance);
	return;
}

sub release_session_lock {
	my ( $self, $instance ) = @_;
	my $lock = 'lock_'.$instance;

	$self->{'memcached'}->delete($lock);

	return;
}

sub retrieve_session {
	my ($self, $instance ) = @_;

	if( ! defined $instance ) {
		return;
	}

	$self->acquire_session_lock( $instance );

	if( my $session = $self->{'memcached'}->get($instance) ) {
		return($session);
	}
	
	return( { '_instance' => $instance } );
}

sub store_session {
	my ($self, $session ) = @_;
	my $instance = $session->{'_instance'};
	my $expire = defined $self->{'memcached_expire'} ? $self->{'memcached_expire'} : 300;

	if( ! defined $session || ! defined $instance ) {
		return;
	}
	
	$self->{'memcached'}->set($instance, $session, $expire);

	$self->release_session_lock($instance);

	return;
}

sub get_virtual_host {
	my $self = shift;
	my $server = $self->{server};
	my $client = $server->{client};
	my $vhost_port;
	my $is_socket = $client && $client->UNIVERSAL::can('NS_proto') &&
           $client->NS_proto eq 'UNIX';

	if( $is_socket ) {
		$vhost_port = Net::Server->VERSION >= 2 ? $client->NS_port
			: $client->NS_unix_path;
	} else {
		$vhost_port = $self->{'server'}->{'sockport'};
	}
	my $vhost = $self->{'virtual_hosts'}->{$vhost_port};
	if( ! defined $vhost ) {
		die('no virtual host defined for port '.$vhost_port);
	}
	return($vhost);
}

sub get_dbh {
	my $self = shift;
	if( ! Mail::MtPolicyd::SqlConnection->is_initialized ) {
		die('no database connection available (no configured?)');
	}
	return( Mail::MtPolicyd::SqlConnection->instance->dbh );
}

sub _is_loglevel {
	my ( $self, $level ) = @_;
	if( $self->{'server'}->{'log_level'} &&
			$self->{'server'}->{'log_level'} >= $level ) {
		return(1);
	}
	return(0);
}

sub _process_one_request {
	my ( $self, $conn, $vhost, $r ) = @_;
	my $port = $vhost->port;
	my $s;
	my $error;

	eval {
		my $start_t = [gettimeofday];
		local $SIG{'ALRM'} = sub { die "Request timeout!" };
		my $timeout = $self->{'request_timeout'};
		alarm($timeout);

		if( $self->_is_loglevel(4) ) { $self->log(4, 'request: '.$r->dump_attr); }
		my $instance = $r->attr('instance');

        Mail::MtPolicyd::Profiler->tick('retrieve session');
		$s = $self->retrieve_session($instance);
		if( $self->_is_loglevel(4) ) { $self->log(4, 'session: '.Dumper($s)); }
		$r->session($s);

        Mail::MtPolicyd::Profiler->tick('run vhost');
		my $result = $vhost->run($r);

		my $response = $result->as_policyd_response;
		$conn->print($response);
		$conn->flush;

		# convert to ms and round by 0.5/int
		my $elapsed = int(tv_interval( $start_t, [gettimeofday] ) * 100 + 0.5);
		$self->log(1, $vhost->name.': instance='.$instance.', type='.$r->type.', t='.$elapsed.'ms, result='.$result->as_log);
	};
	if ( $@ ) { $error = $@; }

	if( defined $s ) {
		$self->store_session($s);
	}

	if( defined $error ) { die( $error ); }

	return;
}

sub process_request {
	my ( $self, $conn ) = @_;
	my $max_keepalive = $self->{'max_keepalive'};

	my $vhost = $self->get_virtual_host;
	my $port = $vhost->port;
	$self->log(4, 'accepted connection on port '.$port );

	for( my $alive_count = 0
			; $max_keepalive == 0 || $alive_count < $max_keepalive 
			; $alive_count++ ) {
		my $r;
		$self->_set_process_stat($vhost->name.', waiting request');
        Mail::MtPolicyd::Profiler->reset;
		eval {
			local $SIG{'ALRM'} = sub { die "Keepalive connection timeout" };
			my $timeout = $self->{'keepalive_timeout'};
			alarm($timeout);
            Mail::MtPolicyd::Profiler->tick('parsing request');
			$r = Mail::MtPolicyd::Request->new_from_fh( $conn, 'server' => $self );
		};
		if ( $@ =~ /Keepalive connection timeout/ ) {
			$self->log(3, '['.$port.']: keepalive timeout: closing connection');
			last;
		} elsif($@ =~ /connection closed by peer/) {
			$self->log(3, '['.$port.']: connection closed by peer');
			last;
		} elsif($@) {
			$self->log(0, '['.$port.']: error while reading request: '.$@);
			last;
		
		}
        Mail::MtPolicyd::Profiler->tick('processing request');
		$self->_set_process_stat($vhost->name.', processing request');
		eval { 
			$self->_process_one_request( $conn, $vhost, $r );
		};
		if ( $@ =~ /Request timeout!/ ) {
			$self->log(1, '['.$port.']: request timed out');
			last;
		} elsif($@) {
			$self->log(0, 'error while processing request: '.$@);
			last;
		}
        Mail::MtPolicyd::Profiler->stop_current_timer;
	    if( $self->_is_loglevel(4) ) {
            $self->log(4, Mail::MtPolicyd::Profiler->to_string);
        }
	}

	$self->log(3, '['.$port.']: closing connection');
	$self->_set_process_stat('idle');

	return;
}

sub _set_process_stat {
	my ( $self, $stat ) = @_;
	$0 = $self->{'program_name'}.' ('.$stat.')'
};

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Mail::MtPolicyd - a modular policy daemon for postfix

=head1 VERSION

version 1.15

=head1 DESCRIPTION

Mail::MtPolicyd is the Net::Server class of the mtpolicyd daemon.

=head2 SYNOPSIS

  use Mail::MtPolicyd;
  Mail::MtPolicyd->run;

=head1 AUTHOR

Markus Benning <ich@markusbenning.de>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by Markus Benning <ich@markusbenning.de>.

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991

=cut
